### ISSUE TITLE
dynamo.optimizations.training.aot_autograd does not trace correct overload

### ISSUE THREAD

**Issue by silvasean** [pytorch/pytorch#90923]
### 🐛 Describe the bug

In the following Python program, `x + 2` gets traced as `torch.ops.aten.add.Tensor` instead of `torch.ops.aten.add.Scalar` which would be more technically correct (and is the op that torch.jit.script gives). I noticed this because in Torch-MLIR we verify that the operands match the schema exactly. Is it possible to have aot_autograd trace the more technically correct overload? Or can you provide guidance about how to correctly emulate the promotion semantics in appropriate generality?

Also, to add onto that, `torch.ops.aten.add.Tensor` is missing an argument. I need special-case code in my importer to handle this (by querying the schema for the default value). One possibility could be that PyTorch provides a utility that normalizes the graph of aten ops to "typecheck" correctly against the schemas of the constituent ops including reifying all default values.

```python
from typing import List

import torch
import torch._dynamo as dynamo
from torch._dynamo.optimizations.training import aot_autograd


def my_backend(gm: torch.fx.GraphModule, example_inputs: List[torch.Tensor]):
    gm.print_readable()
    return gm


@dynamo.optimize(aot_autograd(fw_compiler=my_backend))
def f(x):
    return x + 2


example_inputs = (torch.randn(3, 4),)
f(*example_inputs)
```

Output:
```
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: f32[3, 4]):
        # File: /tmp/repro2.py:15, code: return x + 2
        add: f32[3, 4] = torch.ops.aten.add.Tensor(arg0_1, 2);  arg0_1 = None
        return (add,)     
```

### Versions

PyTorch version: 2.0.0.dev20221213+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux rodete (x86_64)
GCC version: (Debian 12.2.0-3) 12.2.0
Clang version: 14.0.6-2
CMake version: version 3.25.0
Libc version: glibc-2.35

Python version: 3.10.8 (main, Nov  3 2022, 15:17:13) [GCC 12.2.0] (64-bit runtime)
Python platform: Linux-5.19.11-1rodete1-amd64-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.24.0rc2
[pip3] torch==2.0.0.dev20221213+cpu
[pip3] torchvision==0.15.0.dev20221213+cpu
[conda] Could not collect

cc @chauhang @penguinwu @avikchaudhuri @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4 @bdhirsh @bobrenjc93 @aorenste @ezyang @anijain2305 @gmagogsfm @zou3519 @msaroufim @wconstab @soumith @ngimel

**Comment by silvasean** [comment]
I managed to get something working using `torch.ops.prim.NumToTensor.Scalar` to promote the argument, but I'm not sure that is sound in all cases (or that it covers all the possible promotion scenarios). Any advice would be appreciated.

**Comment by ngimel** [comment]
This happens because for a few ops (including `add`) the scalar arguments are getting wrapped in python arg parser, before they are traced  https://github.com/pytorch/pytorch/blob/57e2090e21d7dff1b96424980e6124a9d75bb750/torch/csrc/utils/python_arg_parser.cpp#L85-L98. The tracer faithfully traces the overload that dispatcher sees.

**Comment by eellison** [comment]
Related https://github.com/pytorch/pytorch/issues/86128

**Comment by ezyang** [comment]
My current thinking is we should delete `should_allow_numbers_as_tensors`, sweat out all of the manual Scalar overloads (if they don't exist already), and then call it a day.

**Comment by ezyang** [comment]
If we had partial PrimTorch stuff, where we can lower type promotion but not broadcasting, that would work too, as we would take care of the wrapping back to tensor for you at that point

**Comment by ngimel** [comment]
> My current thinking is we should delete `should_allow_numbers_as_tensors`, sweat out all of the manual Scalar overloads (if they don't exist already), and then call it a day.

We should codegen all the scalar overloads #66491, there just never were sufficiently strong reasons to do it

**Comment by silvasean** [comment]
Based on Ed's example in https://github.com/pytorch/pytorch/issues/86128#issuecomment-1265817027 it looks like it is not possible to have a sound lowering through torch.ops.aten.add.Tensor by simply promoting the scalar to a rank-0 tensor. Is my understanding correct? If so, then this is a soundness issue for us in Torch-MLIR.

**Comment by bdhirsh** [comment]
I still think we should do https://github.com/pytorch/pytorch/issues/66491, but I think that the soundness problem is avoidable. When I print out the graph from compiling this code:
```
torch.compile(backend="aot_eager")
def f(a, b):
    return a * b, a * 2.0

# graph:

class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: i32[2], arg1_1: f64[]):
        # File: tmp7.py:7, code: return a * b, a * 2.0
        mul: f64[2] = torch.ops.aten.mul.Tensor(arg0_1, arg1_1);  arg1_1 = None
        mul_1: f32[2] = torch.ops.aten.mul.Tensor(arg0_1, 2.0);  arg0_1 = None
        return (mul, mul_1)
```

Even though the graph uses the `mul.Tenor` overload, it still bakes in the scalar as an argument into the graph, so the backend has enough info to differentiate:
```
(Pdb) print([x for x in fw_module.graph.nodes][3].format_node())
%mul_1 : [#users=1] = call_function[target=torch.ops.aten.mul.Tensor](args = (%arg0_1, 2.0), kwargs = {})

(Pdb) print(type([x for x in fw_module.graph.nodes][3].args[0]))
<class 'torch.fx.node.Node'>

(Pdb) print(type([x for x in fw_module.graph.nodes][3].args[1]))
<class 'float'>
```

**Comment by silvasean** [comment]
So would the workaround be something like:

1. for all operators identified by Natalia in: https://github.com/pytorch/pytorch/blob/57e2090e21d7dff1b96424980e6124a9d75bb750/torch/csrc/utils/python_arg_parser.cpp#L85-L98
2.  If `isinstance(node.args[1], float)`, then rewrite the `.Tensor` overload to `.Scalar` as we import?

**Comment by bdhirsh** [comment]
That seems reasonable (and would more accurately reflect the end-state that https://github.com/pytorch/pytorch/issues/66491 is supposed to achieve). One problem you might run into though is that not all of those ops have `.Scalar` overloads today. As part of that github issue, we probably want to codegen them.
